rclcpp/src/rclcpp/executors/multi_threaded_executor.cpp

```diff
@@ -85,6 +85,13 @@ MultiThreadedExecutor::run(size_t)
       }
       if (any_exec.timer) {
         // Guard against multiple threads getting the same timer.
+
+        // THIS MUST BE DONE WITH BOTH THE WAIT_MUTEX AND SCHEDULED TIMERS MUTEX TAKEN!!
+        // It might seem unnecessary, but you avoid the following race:
+        // Thread A: get timer, context switch.
+        // Thread B: get timer, insert scheduled timer, execute timer, remove scheduled timer.
+        // Thread A: execute timer.
+        std::lock_guard<std::mutex> scheduled_timers_guard{scheduled_timers_mutex_};
```

Comment:
4. Thread-B resumes with Timer-X, and it just seems not to be scheduled yet. so it will execute the same Timer-X.